### PR TITLE
Add examples for clustering

### PR DESCRIPTION
**Pull request by Ivanidzo4ka**
Address #205

### CODE REVIEW

**Comment by asthana86** [comment]
This is great!, can we also add this as an E2E sample in dotnet/machinelearning/samples with a readme.md similar to the ones we are adding for regression, binary and multi-class classification!

**Comment by Ivanidzo4ka** [comment]
I don't see "dotnet/machinelearning/samples" repo. Can you provide link to it? 

---
In reply to: [391724603](https://github.com/dotnet/machinelearning/pull/222#issuecomment-391724603) [](ancestors = 391724603)

File: test/Microsoft.ML.Tests/Scenarios/ClusteringTests.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    public partial class ScenariosTests
+    {
+        [Fact(Skip = "Missing data set. See https://github.com/dotnet/machinelearning/issues/203")]
+        public void PredictNewsCluster()
+        {
+            string dataPath = GetDataPath(@"external/20newsgroups.txt");
+
+            var pipeline = new LearningPipeline();
+            pipeline.Add(new TextLoader(dataPath).CreateFrom<NewsData>(useHeader: false));
+            pipeline.Add(new CategoricalOneHotVectorizer("Label"));
```
Comment (zeahmed) [inline test/Microsoft.ML.Tests/Scenarios/ClusteringTests.cs]: I think "Label" is not used in clustering unless model is being evaluated against true labels. Why CategoricalOneHotVectorizer is being applied on "Label"?

File: test/Microsoft.ML.Tests/Scenarios/ClusteringTests.cs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+                        (float)(clusters[index].Points[0] + shift),
+                        (float)(clusters[index].Points[1] + shift)
+                    }
+                };
+            }
+            var pipeline = new LearningPipeline();
+            pipeline.Add(CollectionDataSource.Create(data));
+            pipeline.Add(new KMeansPlusPlusClusterer() { K = k });
+            var model = pipeline.Train<ClusteringData, ClusteringPrediction>();
+            //validate no initial centers of clusters belong to same cluster.
```
Comment (zeahmed) [inline test/Microsoft.ML.Tests/Scenarios/ClusteringTests.cs]: Just a minor comment.
`//validate no initial centers of clusters belong to same cluster.`
These don't seem to be `initial center` as these are not set as initial cluster centers to KMean trainer. That is what initial center means in KMean or other clustering algorithms. 

Rather, these are just a data points curated in a way that these appear to be cluster centers initially.

Comment (Ivanidzo4ka) [reply test/Microsoft.ML.Tests/Scenarios/ClusteringTests.cs]: is current phrasing better?

---
In reply to: [190697440](https://github.com/dotnet/machinelearning/pull/222#discussion_r190697440) [](ancestors = 190697440)

File: test/Microsoft.ML.Tests/Scenarios/ClusteringTests.cs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+{
+    public partial class ScenariosTests
+    {
+        [Fact(Skip = "Missing data set. See https://github.com/dotnet/machinelearning/issues/203")]
+        public void PredictNewsCluster()
+        {
+            string dataPath = GetDataPath(@"external/20newsgroups.txt");
+
+            var pipeline = new LearningPipeline();
+            pipeline.Add(new TextLoader(dataPath).CreateFrom<NewsData>(useHeader: false));
```
[the remark this answers is not in the document]
Comment (Ivanidzo4ka) [reply test/Microsoft.ML.Tests/Scenarios/ClusteringTests.cs]: Data set I have is actually have quotes inside mail content. but's it's definitely not sparse

---
In reply to: [190677490](https://github.com/dotnet/machinelearning/pull/222#discussion_r190677490) [](ancestors = 190677490)

File: test/Microsoft.ML.Tests/Scenarios/ClusteringTests.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                TextCase = TextNormalizerTransformCaseNormalizationMode.Lower,
+                StopWordsRemover = new PredefinedStopWordsRemover(),
+                VectorNormalizer = TextTransformTextNormKind.L2,
+                CharFeatureExtractor = new NGramNgramExtractor() { NgramLength = 3, AllLengths = false },
+                WordFeatureExtractor = new NGramNgramExtractor() { NgramLength = 1, AllLengths = true }
+            });
+
+            pipeline.Add(new KMeansPlusPlusClusterer() { K = 20 });
+            var model = pipeline.Train<NewsData, ClusteringPrediction>();
+            var gunResult = model.Predict(new NewsData() { Subject = "Let's disscuss gun control", Content = @"The United States has 88.8 guns per 100 people, or about 270,000,000 guns, which is the highest total and per capita number in the world. 22% of Americans own one or more guns (35% of men and 12% of women). America's pervasive gun culture stems in part from its colonial history, revolutionary roots, frontier expansion, and the Second Amendment, which states: ""A well regulated militia,
```
Comment (TomFinley) [inline test/Microsoft.ML.Tests/Scenarios/ClusteringTests.cs]: Oh good I'm glad we didn't decide to write anything controversial here. 😄

**Comment by justinormont** [comment]
@asthana86 
> This is great!, can we also add this as an E2E sample in dotnet/machinelearning/samples with a readme.md similar to the ones we are adding for regression, binary and multi-class classification!

@zyw400 may have some samples for clustering which we can move to the repo.

**Pull request merged**